devito/passes/iet/misc.py

```diff
@@ -75,48 +73,58 @@ def relax_incr_dimensions(iet, **kwargs):
-        body = compose_nodes(outer)
-        dynamic_parameters = flatten((i.symbolic_bounds, i.step) for i in outer)
-        dynamic_parameters.extend([i.step for i in inner if not is_integer(i.step)])
-        efunc = make_efunc(name, body, dynamic_parameters)
-
-        efuncs.append(efunc)
-
-        # Create the ElementalCalls
-        calls = []
-        for p in product(*ranges):
-            dynamic_args_mapper = {}
-            for i, (m, M, b) in zip(outer, p):
-                dynamic_args_mapper[i.symbolic_min] = m
-                dynamic_args_mapper[i.symbolic_max] = M
-                dynamic_args_mapper[i.step] = b
-                for j in inner:
-                    if j.dim.root is i.dim.root and not is_integer(j.step):
-                        value = j.step if b is i.step else b
-                        dynamic_args_mapper[j.step] = (value,)
-            calls.append(efunc.make_call(dynamic_args_mapper))
-
-        mapper[root] = List(body=calls)
-
-    iet = Transformer(mapper).visit(iet)
+            ranges[i.dim.root] = i.symbolic_max
+
+        # A dictionary to map maximum of parent dimensions
+        # useful for hierarchical blocking
+        proc_parents_max = {}
+
+        # Process inner iterations
+        for n, i in enumerate(inner):
+
+            # Candidate 1: Retrieve parent's max from outer limits
+            parent_max = ranges[i.dim.root]
+
+            # Candidate 2: The symbolic_size of an inner Iteration
+            # may exceed the size of a parent's
+            # block size. Proper margin should be allowed
+            # For cases where upperbound - lowerbound > block_size:
+            # i.symbolic_size > i.dim.parent.step
+            low_margin = i.symbolic_min - i.dim.symbolic_min
+            size_extent = i.symbolic_size - i.dim.parent.step
 
+            upper_ext = i.dim.parent.symbolic_max + size_extent + low_margin
+
+            # Maximum of upper bound candidates
+            it_max = Max(upper_ext, parent_max)
+
+            # In case of hierarchical blocking we should take care not to exceed the
+            # maximum of the parent dimension.
+            # In case parent dim has been processed in the current tree:
+            # if iteration size < parent block_size/step
+            #    upper bound is the parent_max
+            # else/otherwise
+            #    upper bound is the minimum of it_max and parent_max
+
+            if (i.dim.parent in proc_parents_max.keys() and
+               i.symbolic_size <= i.dim.parent.step):
+                upper_bound = Min(proc_parents_max[i.dim.parent])
+            else:
+                upper_bound = Min(i.symbolic_max, it_max)
+
+            # Store selected maximum of this iteration's dimension for
+            # children iteration reference in hierarchical blocking
+            proc_parents_max[i.dim] = upper_bound
+
+            mapper[i] = i._rebuild(limits=(i.symbolic_min, INT(upper_bound), i.step))
```

Comment:
what's this INT(...) for?

Reply:
sure yeah, misunderstanding, I wanted to post sth different,